src/qpx_mmc_css.cpp

```diff
@@ -0,0 +1,1909 @@
+ * p_struct_disckey: the 2048 byte DVD_STRUCT_DISCKEY data
+ * p_disc_key: result, the 5 byte disc key
+ *****************************************************************************/
+static int css_DecryptDiscKey(drive_info *drive,
+                              unsigned char const *p_struct_disckey,
+                              dvd_key_t p_disc_key) {
+  dvd_key_t p_verify;
+  unsigned int i, n = 0;
+
+  /* Decrypt disc key with the above player keys */
+  for (n = 0; n < sizeof(player_keys) / sizeof(dvd_key_t); n++) {
+    css_printkey((char *)("CSS: trying player key "), player_keys[n]);
+
+    for (i = 1; i < 409; i++) {
+      /* Check if player key n is the right key for position i. */
+      css_DecryptKey(0, player_keys[n], p_struct_disckey + 5 * i, p_disc_key);
+
+      /* The first part in the struct_disckey block is the
+       * 'disc key' encrypted with itself.  Using this we
+       * can check if we decrypted the correct key. */
+      css_DecryptKey(0, p_disc_key, p_struct_disckey, p_verify);
+
+      /* If the position / player key pair worked then return. */
+      if (memcmp(p_disc_key, p_verify, DVD_KEY_SIZE) == 0) {
+        return 0;
+      }
+    }
+  }
+
+  /* Have tried all combinations of positions and keys,
+   * and we still didn't succeed. */
+  memset(p_disc_key, 0, DVD_KEY_SIZE);
+  return -1;
+}
+
+static int css_investigate(unsigned char *hash, unsigned char *ckey) {
+  dvd_key_t key;
+  css_DecryptKey(0, ckey, hash, key);
+  return memcmp(key, ckey, DVD_KEY_SIZE);
+}
+
+#define K1TABLEWIDTH 10
+
+static int css_CrackDiscKey(drive_info *drive, unsigned char *p_disc_key) {
+  unsigned char B[5] = {0, 0, 0, 0, 0}; /* Second Stage of mangle cipher */
+  unsigned char C[5] = {0, 0, 0, 0, 0}; /* Output Stage of mangle cipher
+                                         * IntermediateKey */
+  unsigned char k[5] = {0, 0, 0, 0, 0}; /* Mangling cipher key
+                                         * Also output from CSS( C ) */
+  unsigned char out1[5];                /* five first output bytes of LFSR1 */
+  unsigned char out2[5];                /* five first output bytes of LFSR2 */
+  unsigned int lfsr1a;                  /* upper 9 bits of LFSR1 */
+  unsigned int lfsr1b;                  /* lower 8 bits of LFSR1 */
+  unsigned int tmp, tmp2, tmp3, tmp4, tmp5;
+  int i, j;
+  unsigned int nStepA;      /* iterator for LFSR1 start state */
+  unsigned int nStepB;      /* iterator for possible B[0]     */
+  unsigned int nTry;        /* iterator for K[1] possibilities */
+  unsigned int nPossibleK1; /* #of possible K[1] values */
+  unsigned char *K1table;   /* Lookup table for possible K[1] */
+  unsigned int *BigTable;   /* LFSR2 startstate indexed by
+                             * 1,2,5 output byte */
+
+  /*
+   * Prepare tables for hash reversal
+   */
+
+  /* initialize lookup tables for k[1] */
+  K1table = (unsigned char *)malloc(65536 * K1TABLEWIDTH);
+  memset(K1table, 0, 65536 * K1TABLEWIDTH);
```

Comment:
Codacy found an issue: [Either the condition 'K1table==NULL' is redundant or there is possible null pointer dereference: K1table.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)